### PR TITLE
PlantUML Class Diagram

### PR DESCRIPTION
**Pull request by ElijahTheRandom**

### CODE REVIEW

File: doc/uml/Class_Diagram [hunk 1]
```diff
@@ -0,0 +1,49 @@
+@startuml
+class Data{
+-Idenitifer
+-Content
+}
```
Comment (ghost) [inline doc/uml/Class_Diagram]: Spelling should at least be fixed to "identifier". Looking below at `Tag`, perhaps this should be called "name" to be consistent with a "rename()" method?

File: doc/uml/Class_Diagram [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
++deleteData()
++createTag()
++deleteTag()
+}
+
+class Relationship{
+-Relationship(Data, Tag)
++returnTag()
++returnData()
++changeRelationship()
```
Comment (ghost) [inline doc/uml/Class_Diagram]: Get rid of this. We can effectively change a relationship just through create/delete; the relationship doesn't do much other than connect `data` and `tag` so it's better viewed through an immutable light.

File: doc/uml/Class_Diagram [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+-Data[]
+-Relationships[]
++createData()
++deleteData()
++createTag()
++deleteTag()
+}
+
+class Relationship{
+-Relationship(Data, Tag)
```
Comment (ghost) [inline doc/uml/Class_Diagram]: If this is a constructor, it should be public: `+Relatonship(data, tag)`. It might be better to assume constructors at this moment and just place fields with something akin to `data_ptr`, `tag_ptr`, which sort of explicitly tells you that this connects two objects that already exist.

File: doc/uml/Class_Diagram [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
++returnTag()
++returnData()
++changeRelationship()
+}
+
+class Constraint{
+~comparaters()
```
Comment (ghost) [inline doc/uml/Class_Diagram]: The PlantUML documentation indicates this is `package private` which I don't think is relevant. I get what you're trying to represent here I think, but I don't know how to do it myself; I just know this isn't it.

File: doc/uml/Class_Diagram [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+Relationship --* Tag
+Relationship --* Data
+
+Database --o Relationship
+Database --o Data
+Database --o Solver
+Database --o Tag
+
+
+Solver --o Constraint
```
Comment (ghost) [inline doc/uml/Class_Diagram]: Composition states that the child can't exist without the parent. Does the `Constraint` exist without the `Solver`? Is this composition?

I think `Solver` likely has an association with `Relationship` as it checks if they satisfy the constraint.

File: doc/uml/Class_Diagram [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
++sort()
+}
+
+Relationship --* Tag
+Relationship --* Data
+
+Database --o Relationship
+Database --o Data
+Database --o Solver
+Database --o Tag
```
Comment (ghost) [inline doc/uml/Class_Diagram]: I think these go the other way around; `Database o-- Relationship`.

See 
image
The aggregation direction should be toward the `Database`. Should this be composition as well? I don't think we can have any of this stuff without the database: it's sort of a singleton that always exists.

File: doc/uml/Class_Diagram [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+
+
+Solver --o Constraint
+@enduml
```
Comment (jonesz) [inline doc/uml/Class_Diagram:49]: As a general comment, the UML spec says that fields and methods should begin with a lowercase letter and written in whateverCaseThisIsCalled.

File: doc/uml/Class_Diagram [hunk 8]
```diff
@@ -0,0 +1,49 @@
+@startuml
+class Data{
+-Idenitifer
+-Content
+}
+
+class Tag{
+-Identifier
```
Comment (ghost) [inline doc/uml/Class_Diagram]: In general, the UML spec says that field names/methods should be written likeThisInAlternatingCase.

https://www.cs.uct.ac.za/mit_notes/software/htmls/ch05s09.html
> Names of attributes and operations start with a lower-case letter, but may have capital letters to indicate where the spaces would be, if the name would normally have spaces.

**Review by ghost — CHANGES_REQUESTED** [review]
This is a good start; like I said in Discord, there's probably multiple diagrams that can be made that show things at certain granularity (think a diagram with arrows with "connects" describing the relationship in plain English). We could showcase multiplicity on the specific associations (1-to-1, 1-to-N), etc. 

Since this will become more or less an artifact that goes in the appendix of the report, I think it's fair to just work on it incrementally and then at some point generate the images and attach them to the report.

:thumbs_up:
